File: src/app/queue.py
import dataclasses
import os
import threading
import uuid
from pathlib import Path
from typing import Any

import yt_dlp  # type: ignore[import-untyped]
from yt_dlp.utils import DownloadError as YtDownloadError  # type: ignore[attr-defined]

from .config import COOKIES_FILE, DOWNLOADS_DIR, get_cookie_browser_targets
from .utils import _build_ydl_opts, _persist_cookies_from_browser, sanitize_filename

QUEUE_ITEM_STATUSES = ("waiting", "downloading", "completed", "failed", "cancelled")


@dataclasses.dataclass
class QueueItem:
    id: str = ""
    url: str = ""
    filename: str = ""
    fmt: str = "mp3"
    quality: str = "720"
    playlist: str = ""
    output_dir: str = ""
    include_thumbnail: bool = False
    status: str = "waiting"
    progress: float = 0.0
    downloaded_bytes: int = 0
    total_bytes: int = 0
    speed: float = 0.0
    eta: int = 0
    error: str = ""
    output_path: str = ""


class DownloadQueueManager:
    def __init__(self, max_concurrent: int = 2):
        self.items: list[QueueItem] = []
        self._lock: threading.Lock = threading.Lock()
        self._max_concurrent: int = max_concurrent
        self._active_count: int = 0

    def _add_cookies(self, ydl_opts: dict[str, Any]) -> None:  # pyright: ignore[reportExplicitAny]
        if COOKIES_FILE.exists():
            ydl_opts["cookiefile"] = str(COOKIES_FILE)
        else:
            for browser_name, profile in get_cookie_browser_targets():
                try:
                    if profile:
                        ydl_opts["cookiesfrombrowser"] = (browser_name, profile)
                    else:
                        ydl_opts["cookiesfrombrowser"] = (browser_name,)
                    return
                except Exception:  # noqa: BLE001,S112
                    continue

    def add(self, url: str, filename: str, fmt: str, quality: str, playlist: str, output_dir: str, include_thumbnail: bool = False) -> QueueItem:
        item = QueueItem(
            id=str(uuid.uuid4())[:8],
            url=url,
            filename=filename,
            fmt=fmt,
            quality=quality,
            playlist=playlist,
            output_dir=output_dir,
            include_thumbnail=include_thumbnail,
        )
        with self._lock:
            self.items.append(item)
        self._try_process()
        return item

    def remove(self, item_id: str) -> bool:
        with self._lock:
            for i, it in enumerate(self.items):
                if it.id == item_id:
                    if it.status == "downloading":
                        it.status = "cancelled"
                    else:
                        _ = self.items.pop(i)
                    return True
        return False

    def skip(self, item_id: str) -> bool:
        with self._lock:
            for i, it in enumerate(self.items):
                if it.id == item_id:
                    if it.status == "downloading":
                        it.status = "cancelled"
                    else:
                        _ = self.items.pop(i)
                    return True
        return False

    def reorder(self, item_id: str, new_index: int) -> None:
        with self._lock:
            idx = next(i for i, it in enumerate(self.items) if it.id == item_id)
            item = self.items.pop(idx)
            new_index = max(0, min(new_index, len(self.items)))
            self.items.insert(new_index, item)

    def clear_completed(self) -> None:
        with self._lock:
            self.items = [it for it in self.items if it.status in ("waiting", "downloading")]

    def set_max_concurrent(self, n: int) -> None:
        with self._lock:
            self._max_concurrent = max(1, n)

    def get_state(self) -> list[dict[str, Any]]:  # pyright: ignore[reportExplicitAny]
        with self._lock:
            return [dataclasses.asdict(it) for it in self.items]

    def _try_process(self) -> None:
        with self._lock:
            while self._active_count < self._max_concurrent:
                waiting = [it for it in self.items if it.status == "waiting"]
                if not waiting:
                    break
                item = waiting[0]
                item.status = "downloading"
                self._active_count += 1
                t = threading.Thread(target=self._download_worker, args=(item,), daemon=True)
                t.start()

    def _download_worker(self, item: QueueItem) -> None:
        ydl_opts: dict[str, Any] = {}  # pyright: ignore[reportExplicitAny]
        try:
            safe_name = sanitize_filename(item.filename)
            ext = item.fmt

            if item.output_dir:
                base_dir = Path(item.output_dir)
            else:
                base_dir = DOWNLOADS_DIR

            if item.playlist:
                subdir = base_dir / item.playlist
                subdir.mkdir(parents=True, exist_ok=True)
                item.output_path = str(subdir / f"{safe_name}.{ext}")
            else:
                item.output_path = str(base_dir / f"{safe_name}.{ext}")

            if os.path.exists(item.output_path):
                with self._lock:
                    item.status = "completed"
                    item.progress = 100
                return

            ydl_opts = _build_ydl_opts(item.output_path, item.fmt, item.quality, include_thumbnail=item.include_thumbnail)

            def hook(d: dict[str, Any]) -> None:  # pyright: ignore[reportExplicitAny]
                with self._lock:
                    if item.status == "cancelled":
                        raise YtDownloadError("Skipped")
                    item.downloaded_bytes = d.get("downloaded_bytes", 0)
                    item.total_bytes = d.get("total_bytes") or d.get("total_bytes_estimate", 0)
                    item.speed = d.get("speed", 0)
                    item.eta = d.get("eta", 0)
                    if item.total_bytes > 0:
                        item.progress = (item.downloaded_bytes / item.total_bytes) * 100
                    if d.get("status") == "finished" or item.downloaded_bytes >= item.total_bytes > 0:
                        item.progress = 100

            ydl_opts["progress_hooks"] = [hook]
            self._add_cookies(ydl_opts)

            with yt_dlp.YoutubeDL(ydl_opts) as ydl:  # pyright: ignore[reportArgumentType]
                _ = ydl.download([item.url])

            with self._lock:
                item.status = "completed"
                item.progress = 100
        except YtDownloadError as e:
            error_str = str(e)
            if error_str == "Skipped":
                with self._lock:
                    item.status = "cancelled"
                return
            if "cookiefile" in ydl_opts and ("403" in error_str or "forbidden" in error_str.lower()):
                for browser_name, profile in get_cookie_browser_targets():
                    try:
                        retry_opts = {k: v for k, v in ydl_opts.items() if k != "cookiefile"}
                        if profile:
                            retry_opts["cookiesfrombrowser"] = (browser_name, profile)
                        else:
                            retry_opts["cookiesfrombrowser"] = (browser_name,)
                        with yt_dlp.YoutubeDL(retry_opts) as ydl:  # pyright: ignore[reportArgumentType]
                            _ = ydl.download([item.url])
                        _persist_cookies_from_browser(browser_name, profile)
                        with self._lock:
                            item.status = "completed"
                            item.progress = 100
                        return
                    except YtDownloadError:
                        continue
            with self._lock:
                item.status = "failed"
                item.error = error_str
        except Exception as e:  # noqa: BLE001
            with self._lock:
                item.status = "failed"
                item.error = str(e)
        finally:
            with self._lock:
                self._active_count -= 1
            self._try_process()


download_queue = DownloadQueueManager(max_concurrent=2)

File: src/lib/queue.ts
import type { QueueItemData } from "@/types";
import {
  sanitizeFilename,
  parseProgressLine,
  getCookieBrowserTargets,
} from "@/lib/utils";
import { COOKIES_FILE, DOWNLOADS_DIR } from "@/config/paths";
import { extractVideoId, generatePoToken, trySpawnYtdlp } from "@/lib/ytdlp";
import { unlinkSync } from "fs";
import PQueue from "p-queue";
import Logger from "@/lib/logger";

const logger = new Logger("QUEUE");

const PROGRESS_TEMPLATE =
  "download:[download] %(progress.downloaded_bytes)s %(progress.total_bytes)s %(progress._percent_str)s %(progress._speed_str)s %(progress._eta_str)s %(progress.speed)s %(progress.eta)s";

class DownloadQueueManager {
  items: QueueItemData[] = [];
  private queue: PQueue;
  private itemIdCounter = 0;

  constructor(maxConcurrent = 2) {
    this.queue = new PQueue({ concurrency: maxConcurrent });
  }

  private genId(): string {
    return (++this.itemIdCounter).toString(36).padStart(6, "0");
  }

  get maxConcurrent(): number {
    return this.queue.concurrency;
  }

  set maxConcurrent(n: number) {
    this.queue.concurrency = Math.max(0, n);
  }

  add(
    params: {
      url: string;
      filename: string;
      fmt: string;
      quality: string;
      playlist: string;
      output_dir: string;
      include_thumbnail: boolean;
    },
    priority = 0,
  ): string {
    const item: QueueItemData = {
      id: this.genId(),
      url: params.url,
      filename: params.filename,
      fmt: params.fmt,
      quality: params.quality,
      playlist: params.playlist,
      output_dir: params.output_dir,
      include_thumbnail: params.include_thumbnail,
      status: "waiting",
      progress: 0,
      downloaded_bytes: 0,
      total_bytes: 0,
      speed: 0,
      eta: 0,
      error: "",
      output_path: "",
    };
    this.items.push(item);
    this.queue.add(() => this._processItem(item), { priority });
    logger.log(
      `add id=${item.id} "${item.filename}" fmt=${item.fmt} quality=${item.quality} priority=${priority}`,
    );
    return item.id;
  }

  remove(itemId: string): boolean {
    const idx = this.items.findIndex((it) => it.id === itemId);
    if (idx === -1) return false;
    const item = this.items[idx]!;
    if (item.status === "downloading") {
      item.status = "cancelled";
    } else {
      this.items.splice(idx, 1);
    }
    logger.log(`remove id=${itemId} status=${item.status}`);
    return true;
  }

  skip(itemId: string): boolean {
    const ok = this.remove(itemId);
    if (ok) logger.log(`skip id=${itemId}`);
    return ok;
  }

  reorder(itemId: string, newIndex: number): void {
    const idx = this.items.findIndex((it) => it.id === itemId);
    if (idx === -1) return;
    const [item] = this.items.splice(idx, 1);
    const clamped = Math.max(0, Math.min(newIndex, this.items.length));
    this.items.splice(clamped, 0, item!);
    logger.log(`reorder id=${itemId} from=${idx} to=${clamped}`);
  }

  clearCompleted(): void {
    const before = this.items.length;
    this.items = this.items.filter(
      (it) => it.status === "waiting" || it.status === "downloading",
    );
    const removed = before - this.items.length;
    logger.log(`cleared ${removed} completed items`);
  }

  resolveConflict(itemId: string, action: "overwrite" | "skip"): boolean {
    const item = this.items.find((it) => it.id === itemId);
    if (!item || item.status !== "conflict") return false;
    logger.log(`resolve id=${itemId} action=${action}`);
    if (action === "overwrite") {
      try {
        if (item.output_path) unlinkSync(item.output_path);
      } catch {
        return false;
      }
      item.status = "waiting";
      item.progress = 0;
      this.queue.add(() => this._processItem(item));
    } else {
      item.status = "completed";
      item.progress = 100;
    }
    return true;
  }

  setMaxConcurrent(n: number): void {
    this.maxConcurrent = Math.max(0, n);
    logger.log(`config max_concurrent=${this.maxConcurrent}`);
  }

  getState(): QueueItemData[] {
    return this.items.map((it) => ({ ...it }));
  }

  private async _runWithCookies(
    item: QueueItemData,
    opts: {
      browser?: string;
      profile?: string | null;
      poToken?: string;
      playerClient?: string;
    },
  ): Promise<boolean> {
    if (item.status === "cancelled") return false;
    try {
      const ext = item.fmt === "mp3" ? "mp3" : "mp4";
      const outtmpl = item.output_path.replace(`.${ext}`, ".%(ext)s");
      const args: string[] = [
        item.url,
        "-o",
        outtmpl,
        "--newline",
        "--progress-template",
        PROGRESS_TEMPLATE,
        "--embed-metadata",
      ];

      if (opts.browser) {
        args.push(
          "--cookies-from-browser",
          opts.profile ? `${opts.browser}:${opts.profile}` : opts.browser,
        );
      } else if (await Bun.file(COOKIES_FILE).exists()) {
        args.push("--cookies", COOKIES_FILE);
      }

      if (item.fmt === "mp3") {
        args.push(
          "-f",
          "bestaudio/best",
          "-x",
          "--audio-format",
          "mp3",
          "--audio-quality",
          "0",
        );
      } else {
        args.push(
          "-f",
          `bestvideo[height<=${item.quality}]+bestaudio/best[height<=${item.quality}]`,
          "--merge-output-format",
          "mp4",
        );
      }

      const proc = trySpawnYtdlp(args, {
        poToken: opts.poToken,
        playerClient: opts.playerClient,
      });
      if (!proc) return false;

      const stderrReader = (
        proc.stderr as ReadableStream<Uint8Array>
      ).getReader();
      const decoder = new TextDecoder();
      let stderrText = "";

      try {
        while (true) {
          const { done, value } = await stderrReader.read();
          if (done) break;
          const text = decoder.decode(value, { stream: true });
          stderrText += text;
          for (const line of text.split("\n")) {
            const parsed = parseProgressLine(line);
            if (parsed) {
              if (parsed.downloaded_bytes !== undefined)
                item.downloaded_bytes = parsed.downloaded_bytes;
              if (parsed.total_bytes !== undefined)
                item.total_bytes = parsed.total_bytes;
              if (parsed.speed_bytes !== undefined)
                item.speed = parsed.speed_bytes;
              if (parsed.eta_seconds !== undefined) item.eta = parsed.eta_seconds;
              item.progress = parsed.percent;
            }
          }
        }
      } finally {
        stderrReader.releaseLock();
      }

      const exitCode = await proc.exited;
      if (exitCode === 0) {
        item.status = "completed";
        item.progress = 100;
        logger.log(`complete id=${item.id} (retry) path="${item.output_path}"`);
        return true;
      }
      logger.warn(`retry fail id=${item.id} exit=${exitCode}: ${stderrText.slice(0, 200)}`);
      return false;
    } catch (e) {
      logger.warn(`retry error id=${item.id}: ${e}`);
      return false;
    }
  }

  private async _processItem(item: QueueItemData): Promise<void> {
    if (item.status !== "waiting") return;

    try {
      const safeName = sanitizeFilename(item.filename);
      const ext = item.fmt === "mp3" ? "mp3" : "mp4";

      const baseDir = item.output_dir || DOWNLOADS_DIR;
      const subdir = item.playlist ? `${baseDir}/${item.playlist}` : baseDir;

      item.output_path = `${subdir}/${safeName}.${ext}`;

      item.status = "downloading";
      logger.log(`start id=${item.id} "${safeName}.${ext}"`);

      if (await Bun.file(item.output_path).exists()) {
        item.status = "conflict";
        item.progress = 0;
        logger.log(`conflict id=${item.id} path="${item.output_path}"`);
        return;
      }

      const outtmpl = item.output_path.replace(`.${ext}`, ".%(ext)s");
      const args: string[] = [
        item.url,
        "-o",
        outtmpl,
        "--newline",
        "--progress-template",
        PROGRESS_TEMPLATE,
        "--embed-metadata",
      ];

      if (await Bun.file(COOKIES_FILE).exists()) {
        args.push("--cookies", COOKIES_FILE);
      }

      if (item.include_thumbnail) {
        args.push("--write-thumbnail", "--embed-thumbnail");
      }

      if (item.fmt === "mp3") {
        args.push(
          "-f",
          "bestaudio/best",
          "-x",
          "--audio-format",
          "mp3",
          "--audio-quality",
          "0",
        );
      } else {
        args.push(
          "-f",
          `bestvideo[height<=${item.quality}]+bestaudio/best[height<=${item.quality}]`,
          "--merge-output-format",
          "mp4",
        );
      }

      const proc = trySpawnYtdlp(args);
      if (!proc) throw new Error("yt-dlp not found or failed to start");

      const stderrReader = (
        proc.stderr as ReadableStream<Uint8Array>
      ).getReader();
      const decoder = new TextDecoder();
      let stderrText = "";

      try {
        while (true) {
          const { done, value } = await stderrReader.read();
          if (done) break;
          const text = decoder.decode(value, { stream: true });
          stderrText += text;
          for (const line of text.split("\n")) {
            const parsed = parseProgressLine(line);
            if (parsed) {
              if (parsed.downloaded_bytes !== undefined)
                item.downloaded_bytes = parsed.downloaded_bytes;
              if (parsed.total_bytes !== undefined)
                item.total_bytes = parsed.total_bytes;
              if (parsed.speed_bytes !== undefined)
                item.speed = parsed.speed_bytes;
              if (parsed.eta_seconds !== undefined) item.eta = parsed.eta_seconds;
              item.progress = parsed.percent;
            }
          }
        }
      } finally {
        stderrReader.releaseLock();
      }

      const exitCode = await proc.exited;
      if (exitCode === 0) {
        item.status = "completed";
        item.progress = 100;
        logger.log(`complete id=${item.id} path="${item.output_path}"`);
      } else {
        if (item.status === "cancelled") return;
        logger.warn(
          `fail id=${item.id} exit=${exitCode}: ${stderrText.slice(0, 300)}`,
        );

        const lower = stderrText.toLowerCase();
        const botCheck =
          /sign in to confirm|sign in to continue|not a bot|captcha|exponential backoff|403|forbidden/.test(
            lower,
          );

        if (botCheck) {
          logger.log(`bot-check/sign-in detected for id=${item.id}, retrying`);

          // 1) live browser cookies (default client)
          for (const [browser, profile] of getCookieBrowserTargets()) {
            if (item.status === "cancelled") return;
            if (await this._runWithCookies(item, { browser, profile })) return;
          }

          // 2) saved cookies + PO token + tv client fallback
          if (item.status !== "cancelled") {
            const videoId = extractVideoId(item.url);
            const po = videoId ? await generatePoToken(videoId) : null;
            if (po) {
              const ok = await this._runWithCookies(item, {
                poToken: po,
                playerClient: "tv",
              });
              if (ok) return;
            }
          }

          if (item.status === "cancelled") return;
          item.status = "failed";
          item.error = stderrText || "Download failed (sign-in / bot check)";
          return;
        }

        item.status = "failed";
        item.error = stderrText || "Download failed";
      }
    } catch (e) {
      logger.error(`error id=${item.id}: ${e}`);
      if (item.status === "cancelled") return;
      item.status = "failed";
      item.error = String(e);
    }
  }
}

export const downloadQueue = new DownloadQueueManager(2);
